In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from matplotlib import pyplot
import feather
import warnings
from IPython.display import display
warnings.filterwarnings('ignore')

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [3]:
hv = pd.read_feather("hv.feather")

In [4]:
hv.shape

(1745325, 25)

In [5]:
hv.head()

Unnamed: 0 Invoice/Item Number        Date  Store Number  \
0           2     INV-24107200133  12/23/2019          2508   
1           3     INV-24090200052  12/20/2019          2649   
2           6     INV-24117700068  12/23/2019          2567   
3           7     INV-24105000057  12/23/2019          2538   
4          12     INV-24117300020  12/23/2019          2614   

                               Store Name                    Address  \
0     Hy-Vee Food Store #1 / Cedar Rapids  1843 Johnson Avenue, N.W.   
1                     Hy-Vee #3 / Dubuque              400 Locust St   
2            Hy-Vee Drugstore / Davenport         2200 West Kimberly   
3         Hy-Vee Food Store #3 / Waterloo           1422 Flammang Dr   
4  Hy-Vee #3 Food & Drugstore / Davenport         1823 E Kimberly Rd   

           City  Zip Code                        Store Location  \
0  Cedar Rapids   52405.0           POINT (-91.697941 41.97447)   
1       Dubuque   52001.0  POINT (-90.666497 42.49721900000001)   
2     Davenport   52806.0          POINT (-90.608201 41.560663)   
3      Waterloo   50702.0          POINT (-92.327917 42.459938)   
4     Davenport   52807.0  POINT (-90.548919 41.55678100000001)   

   County Number      County   Category              Category Name  \
0           57.0        LINN  1041100.0          American Dry Gins   
1           31.0     DUBUQUE  1011400.0         Tennessee Whiskies   
2           82.0       SCOTT  1011200.0  Straight Bourbon Whiskies   
3            7.0  BLACK HAWK  1011100.0           Blended Whiskies   
4           82.0       SCOTT  1011100.0           Blended Whiskies   

   Vendor Number         Vendor Name  Item Number  \
0          370.0   PERNOD RICARD USA        32236   
1           85.0  Brown Forman Corp.        26826   
2           65.0     Jim Beam Brands        19066   
3           65.0     Jim Beam Brands        24456   
4           65.0     Jim Beam Brands        27680   

                  Item Description  Pack  Bottle Volume (ml)  \
0           Seagrams Extra Dry Gin  12.0               750.0   
1  Jack Daniels Old #7 Black Label  12.0               750.0   
2                         Jim Beam  12.0               750.0   
3            Kessler Blend Whiskey  12.0               750.0   
4      Jim Beam Kentucky Fire Mini  12.0                50.0   

   State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
0               6.49                 9.74          12.0          116.88   
1              15.57                23.36          12.0          280.32   
2              10.50                15.75           6.0           94.50   
3               5.51                 8.27           3.0           24.81   
4               5.40                 8.10          12.0           97.20   

   Volume Sold (Liters)  Volume Sold (Gallons)  
0                  9.00                   2.37  
1                  9.00                   2.37  
2                  4.50                   1.18  
3                  2.25                   0.59  
4                  0.60                   0.15

# Creating Store category column

In [6]:
hv["Store Category"] = "Supermarket"
hv.loc[hv["Store Name"].str.contains('Food Store'), "Store Category"] = 'Food Store'

In [7]:
hv.loc[hv["Store Name"].str.contains("Wine & Spirits"), "Store Category"] = 'Wine & Spirits'
hv.loc[hv["Store Name"].str.contains("Wine and Spirits"), "Store Category"] = 'Wine & Spirits'

In [8]:
hv.loc[hv["Store Name"].str.contains("Food & Drugstore"), "Store Category"] = "Food & Drugstore"
hv.loc[hv["Store Name"].str.contains("Food and Drug"), "Store Category"] = "Food & Drugstore"
hv.loc[hv["Store Name"].str.contains("Food and Drugstore"), "Store Category"] = "Food & Drugstore"

In [9]:
hv.loc[hv["Store Name"].str.contains("Gas"), "Store Category"] = "Gas"

In [10]:
hv.loc[hv["Store Name"].str.contains("Fast & Fresh"), "Store Category"] = "Fast & Fresh"
hv.loc[hv["Store Name"].str.contains("Fast and Fresh"), "Store Category"] = "Fast & Fresh"

In [11]:
hv.loc[hv["Store Name"].str.contains("Drugstore"), "Store Category"] = "Drugstore"

In [12]:
floattoint = ["Zip Code", "County Number", "Category", "Vendor Number", "Item Number", "Pack", "Bottles Sold"]

for i in floattoint:
    hv[i] = hv[i].astype("Int64")

In [13]:
hv = hv.drop(columns=["Unnamed: 0"])

# Missing Values

In [14]:
null_counts = hv.isnull().sum()
null_counts

Invoice/Item Number           0
Date                          0
Store Number                  0
Store Name                    0
Address                    1503
City                       1503
Zip Code                   1503
Store Location           266054
County Number              1503
County                     1503
Category                    418
Category Name               420
Vendor Number                 1
Vendor Name                   1
Item Number                   0
Item Description              0
Pack                          1
Bottle Volume (ml)            1
State Bottle Cost             1
State Bottle Retail           1
Bottles Sold                  1
Sale (Dollars)                1
Volume Sold (Liters)          1
Volume Sold (Gallons)         1
Store Category                0
dtype: int64

## Missing Category and Category Name

In [15]:
hv.loc[hv["Category"].isna()].head()

Invoice/Item Number        Date  Store Number  \
19       INV-24110700158  12/23/2019          2633   
232      INV-24111100086  12/23/2019          2527   
927      INV-24105000123  12/23/2019          2538   
982      INV-24109000198  12/23/2019          2572   
1583     INV-24115000095  12/23/2019          2619   

                             Store Name               Address  \
19         Hy-Vee #3 / BDI / Des Moines       3221 SE 14th St   
232   Hy-Vee Food Store #5 / Des Moines  1107 SE Army Post Rd   
927     Hy-Vee Food Store #3 / Waterloo      1422 Flammang Dr   
982     Hy-Vee Food Store / Cedar Falls       6301 University   
1583      Hy-Vee Wine and Spirits / WDM         1725  74th St   

                 City  Zip Code                Store Location  County Number  \
19         Des Moines     50320  POINT (-93.596754 41.554101)             77   
232        Des Moines     50315  POINT (-93.602372 41.526289)             77   
927          Waterloo     50702  POINT (-92.327917 42.459938)              7   
982       Cedar Falls     50613  POINT (-92.435236 42.512789)              7   
1583  West Des Moines     50266  POINT (-93.808855 41.598515)             77   

          County  Category Category Name  Vendor Number Vendor Name  \
19          POLK      <NA>          None            305     Mhw LTD   
232         POLK      <NA>          None            305     Mhw LTD   
927   BLACK HAWK      <NA>          None            305     Mhw LTD   
982   BLACK HAWK      <NA>          None            305     Mhw LTD   
1583        POLK      <NA>          None            305     Mhw LTD   

      Item Number                 Item Description  Pack  Bottle Volume (ml)  \
19          80024  Blue Chair Bay Banana Rum Cream    12               750.0   
232         80024  Blue Chair Bay Banana Rum Cream    12               750.0   
927         80024  Blue Chair Bay Banana Rum Cream    12               750.0   
982         80024  Blue Chair Bay Banana Rum Cream    12               750.0   
1583        80024  Blue Chair Bay Banana Rum Cream    12               750.0   

      State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
19                 9.42                14.13            12          169.56   
232                9.42                14.13            12          169.56   
927                9.42                14.13             3           42.39   
982                9.42                14.13             3           42.39   
1583               9.42                14.13             3           42.39   

      Volume Sold (Liters)  Volume Sold (Gallons)  Store Category  
19                    9.00                   2.37     Supermarket  
232                   9.00                   2.37      Food Store  
927                   2.25                   0.59      Food Store  
982                   2.25                   0.59      Food Store  
1583                  2.25                   0.59  Wine & Spirits

In [16]:
missing_category_item_numbers = hv[hv["Category"].isna()]["Item Number"].dropna().unique()

In [ ]:
missing_category_name_item_numbers = hv[hv["Category Name"].isna()]["Item Number"].dropna().unique()

In [17]:
for item_number in missing_category_item_numbers:
    # Grab mode for `Category` using `Item Number`
    category_from_mode = hv.loc[(hv["Item Number"] == item_number)]["Category"].mode()

    # For now skipping none mode values
    if len(category_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    category_from_mode = category_from_mode[0]
    
    # Find where `Category` is missing ("N/A" or "None")
    #    Then replace with new `Category` from mode
    hv.loc[hv['Item Number'] == item_number, 'Category'] = hv.loc[hv['Item Number'] == item_number, 'Category'].fillna(category_from_mode)
    
for item_number in missing_category_name_item_numbers:
    # Grab mode for `Category Name` using `Item Number`
    category_name_from_mode = hv.loc[(hv["Item Number"] == item_number)]["Category Name"].mode()

    # For now skipping none mode values
    if len(category_name_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    category_name_from_mode = category_name_from_mode[0]
    
    # Find where `Category Name` is missing ("N/A" or "None")
    #    Then replace with new `Category Name` from mode
    hv.loc[hv['Item Number'] == item_number, 'Category Name'] = hv.loc[hv['Item Number'] == item_number, 'Category Name'].fillna(category_name_from_mode)

In [18]:
# Find all item numbers of null Categories
new_missing_category_item_numbers = hv[hv['Category'].isna()]['Item Number'].unique()
# Find all item numbers of null Categories Names
new_missing_category_name_item_numbers = hv[hv['Category Name'].isna()]['Item Number'].unique()

# New missing items should only have null modes
# meaning all same items in dataset are null (we can't set them to similar items)
for item_number in new_missing_category_item_numbers:
    category_from_mode = hv.loc[(hv["Item Number"] == item_number)]["Category"].mode()
    display(category_from_mode)
    
for item_number in new_missing_category_name_item_numbers:
    category_name_from_mode = hv.loc[(hv["Item Number"] == item_number)]["Category Name"].mode()
    display(category_name_from_mode)

Series([], dtype: Int64)

Series([], dtype: Int64)

Series([], dtype: Int64)

Series([], dtype: Int64)

Series([], dtype: Int64)

Series([], dtype: Int64)

Series([], dtype: object)

Series([], dtype: object)

Series([], dtype: object)

Series([], dtype: object)

Series([], dtype: object)

Series([], dtype: object)

Series([], dtype: object)

Series([], dtype: object)

In [19]:
hv.loc[hv["Category"].isna()]

Invoice/Item Number        Date  Store Number  \
54848       INV-24573900003  01/15/2020          2625   
1172920     INV-20382200001  07/03/2019          2648   
1291094     INV-25817300011  03/11/2020          2673   
1292514     INV-25844800003  03/12/2020          2626   
1295098     INV-25824400019  03/12/2020          2675   
1436349     INV-26611600001  04/16/2020          2515   
1483107     INV-01622300001  11/16/2016          2556   
1562201     INV-27171500042  05/11/2020          2662   
1580086     INV-27345600003  05/19/2020          2538   

                                    Store Name                Address  \
54848     Hy-Vee Wine & Spirits #2 / Davenport     3301 W Kimberly Rd   
1172920                        Hy-Vee #4 / WDM          555 S 51st St   
1291094                 Hy-Vee # 6/ Des Moines          420 Court Ave   
1292514    Hy-Vee Drugstore / University / DSM    4100 University Ave   
1295098                 Hy-Vee #2 / Coralville      3285 Crosspark Rd   
1436349      Hy-Vee Food Store #1 / Mason City         2400 4th St SW   
1483107  Hy-Vee Wine and Spirits / Estherville                   None   
1562201      Hy-Vee Wine & Spirits / Muscatine  522 Mulberry, Suite A   
1580086        Hy-Vee Food Store #3 / Waterloo       1422 Flammang Dr   

                    City  Zip Code                        Store Location  \
54848          Davenport     52804                                  None   
1172920  West Des Moines     50265          POINT (-93.773557 41.561197)   
1291094       Des Moines     50309  POINT (-93.62245400000002 41.584842)   
1292514       Des Moines     50311          POINT (-93.673222 41.600361)   
1295098       Coralville     52241  POINT (-91.60627100000002 41.721438)   
1436349       Mason City     50401          POINT (-93.235588 43.148451)   
1483107             None      <NA>                                  None   
1562201        Muscatine     52761           POINT (-91.04437 41.426615)   
1580086         Waterloo     50702          POINT (-92.327917 42.459938)   

         County Number      County  Category Category Name  Vendor Number  \
54848               82       SCOTT      <NA>          None            205   
1172920             77        POLK      <NA>          None            322   
1291094             77        POLK      <NA>          None            205   
1292514             77        POLK      <NA>          None            205   
1295098             52     JOHNSON      <NA>          None            205   
1436349             17  CERRO GORD      <NA>          None             65   
1483107           <NA>        None      <NA>          None            260   
1562201             70   MUSCATINE      <NA>          None            125   
1580086              7  BLACK HAWK      <NA>          None            125   

                           Vendor Name  Item Number  \
54848               E & J Gallo Winery       994771   
1172920  Prestige Wine & Spirits Group       994811   
1291094             E & J Gallo Winery       994771   
1292514             E & J Gallo Winery       994771   
1295098             E & J Gallo Winery       994771   
1436349                Jim Beam Brands       973508   
1483107                DIAGEO AMERICAS          115   
1562201       CEDAR RIDGE VINEYARDS LL        27849   
1580086       CEDAR RIDGE VINEYARDS LL       926786   

                                          Item Description  Pack  \
54848                   SOOH Diplomatico Reserva Exclusiva     6   
1172920                                    Athena Ouzo Rum    12   
1291094                      Diplomatico Reserva Exclusiva     6   
1292514                      Diplomatico Reserva Exclusiva     6   
1295098                      Diplomatico Reserva Exclusiva     6   
1436349                         Dekuyper Mixologist Ginger    12   
1483107                      Crown Royal w/Glass & Coaster     6   
1562201                        Cedar Ridge Reserve Bourbon     6   
1580086  Ce

In [20]:
null_counts = hv.isnull().sum()
null_counts

Invoice/Item Number           0
Date                          0
Store Number                  0
Store Name                    0
Address                    1503
City                       1503
Zip Code                   1503
Store Location           266054
County Number              1503
County                     1503
Category                      9
Category Name                11
Vendor Number                 1
Vendor Name                   1
Item Number                   0
Item Description              0
Pack                          1
Bottle Volume (ml)            1
State Bottle Cost             1
State Bottle Retail           1
Bottles Sold                  1
Sale (Dollars)                1
Volume Sold (Liters)          1
Volume Sold (Gallons)         1
Store Category                0
dtype: int64

### Diplomatico Reserva Exclusiva and SOOH Diplomatico Reserva Exclusiva

In [21]:
hv.loc[(hv["Item Description"] == "Diplomatico Reserva Exclusiva") & (hv["Category Name"].notna())][["Item Description", "Category", "Category Name"]].head()

Item Description  Category  Category Name
92170   Diplomatico Reserva Exclusiva   1062300  Aged Dark Rum
93315   Diplomatico Reserva Exclusiva   1062300  Aged Dark Rum
116260  Diplomatico Reserva Exclusiva   1062300  Aged Dark Rum
124886  Diplomatico Reserva Exclusiva   1062300  Aged Dark Rum
129055  Diplomatico Reserva Exclusiva   1062300  Aged Dark Rum

In [22]:
#only one transaction for SOOH Diplomatico Reserva Exclusiva, 
# will give category and cat name same as Diplomatico Reserva Exclusiva

hv.loc[(hv["Item Description"] == "SOOH Diplomatico Reserva Exclusiva")]

Invoice/Item Number        Date  Store Number  \
54848     INV-24573900003  01/15/2020          2625   

                                 Store Name             Address       City  \
54848  Hy-Vee Wine & Spirits #2 / Davenport  3301 W Kimberly Rd  Davenport   

       Zip Code Store Location  County Number County  Category Category Name  \
54848     52804           None             82  SCOTT      <NA>          None   

       Vendor Number         Vendor Name  Item Number  \
54848            205  E & J Gallo Winery       994771   

                         Item Description  Pack  Bottle Volume (ml)  \
54848  SOOH Diplomatico Reserva Exclusiva     6               750.0   

       State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
54848              19.33                 29.0             6           174.0   

       Volume Sold (Liters)  Volume Sold (Gallons)  Store Category  
54848                   4.5                   1.18  Wine & Spirits

In [23]:
hv.loc[hv['Item Description'] == "Diplomatico Reserva Exclusiva", 'Category'] = 1062300
hv.loc[hv['Item Description'] == "Diplomatico Reserva Exclusiva", 'Category Name'] = "Aged Dark Rum"

hv.loc[(hv["Item Description"] == "Diplomatico Reserva Exclusiva") & (hv["Category"].isna())][["Item Description","Category", "Category Name"]]

Empty DataFrame
Columns: [Item Description, Category, Category Name]
Index: []

In [24]:
hv.loc[hv['Item Description'] == "SOOH Diplomatico Reserva Exclusiva", 'Category'] = 1062300
hv.loc[hv['Item Description'] == "SOOH Diplomatico Reserva Exclusiva", 'Category Name'] = "Aged Dark Rum"

hv.loc[(hv["Item Description"] == "SOOH Diplomatico Reserva Exclusiva") & (hv["Category"].isna())][["Item Description","Category", "Category Name"]]

Empty DataFrame
Columns: [Item Description, Category, Category Name]
Index: []

### Athena Ouzo Rum

In [25]:
hv.loc[(hv["Item Description"] == "Athena Ouzo Rum")]

Invoice/Item Number        Date  Store Number       Store Name  \
1172920     INV-20382200001  07/03/2019          2648  Hy-Vee #4 / WDM   

               Address             City  Zip Code  \
1172920  555 S 51st St  West Des Moines     50265   

                       Store Location  County Number County  Category  \
1172920  POINT (-93.773557 41.561197)             77   POLK      <NA>   

        Category Name  Vendor Number                    Vendor Name  \
1172920          None            322  Prestige Wine & Spirits Group   

         Item Number Item Description  Pack  Bottle Volume (ml)  \
1172920       994811  Athena Ouzo Rum    12               750.0   

         State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
1172920               6.06                 9.09            12          109.08   

         Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
1172920                   9.0                   2.37    Supermarket

In [26]:
hv.loc[(hv["Item Number"] == 994811)]

Invoice/Item Number        Date  Store Number       Store Name  \
1172920     INV-20382200001  07/03/2019          2648  Hy-Vee #4 / WDM   

               Address             City  Zip Code  \
1172920  555 S 51st St  West Des Moines     50265   

                       Store Location  County Number County  Category  \
1172920  POINT (-93.773557 41.561197)             77   POLK      <NA>   

        Category Name  Vendor Number                    Vendor Name  \
1172920          None            322  Prestige Wine & Spirits Group   

         Item Number Item Description  Pack  Bottle Volume (ml)  \
1172920       994811  Athena Ouzo Rum    12               750.0   

         State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
1172920               6.06                 9.09            12          109.08   

         Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
1172920                   9.0                   2.37    Supermarket

#### Athena Ouzo Rum is missing a category name and number
- Filtering item description by those that contain "Rum" and also filter by County POLK
- Looking online, Athena Ouzo spirit style is Herbal & Spice rum
- Setting Category Name to Spiced Rum and using the most common Category in POLK county for Rum Category

In [27]:
hv.loc[(hv["Item Description"] == "Athena Ouzo Rum") & (hv["Category Name"].notna())][["Category", "Category Name"]].head()

Empty DataFrame
Columns: [Category, Category Name]
Index: []

In [28]:
hv[(hv['Item Description'].str.contains('Rum', regex=False)) & (hv["County"] == "POLK")].head()

Invoice/Item Number        Date  Store Number  \
19      INV-24110700158  12/23/2019          2633   
69      INV-24110700091  12/23/2019          2633   
232     INV-24111100086  12/23/2019          2527   
413     INV-24110400015  12/23/2019          2680   
523     INV-24115000015  12/23/2019          2619   

                            Store Name                    Address  \
19        Hy-Vee #3 / BDI / Des Moines            3221 SE 14th St   
69        Hy-Vee #3 / BDI / Des Moines            3221 SE 14th St   
232  Hy-Vee Food Store #5 / Des Moines       1107 SE Army Post Rd   
413          Hy-Vee Fulfillment Center  3303 86th Street, Suite B   
523      Hy-Vee Wine and Spirits / WDM              1725  74th St   

                City  Zip Code                Store Location  County Number  \
19        Des Moines     50320  POINT (-93.596754 41.554101)             77   
69        Des Moines     50320  POINT (-93.596754 41.554101)             77   
232       Des Moines     50315  POINT (-93.602372 41.526289)             77   
413        Urbandale     50322  POINT (-93.736458 41.624917)             77   
523  West Des Moines     50266  POINT (-93.808855 41.598515)             77   

    County  Category   Category Name  Vendor Number         Vendor Name  \
19    POLK   1062500    Flavored Rum            305             Mhw LTD   
69    POLK   1062200       White Rum            380   Phillips Beverage   
232   POLK   1062500    Flavored Rum            305             Mhw LTD   
413   POLK   1062400      Spiced Rum            259  Heaven Hill Brands   
523   POLK   1081200  Cream Liqueurs            305             Mhw LTD   

     Item Number                 Item Description  Pack  Bottle Volume (ml)  \
19         80024  Blue Chair Bay Banana Rum Cream    12               750.0   
69         45419               Phillips White Rum    12              1000.0   
232        80024  Blue Chair Bay Banana Rum Cream    12               750.0   
413        43028        Admiral Nelson Spiced Rum     6              1750.0   
523        73051                    Rumchata Mini    12                50.0   

     State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
19                9.42                14.13            12          169.56   
69                4.45                 6.68            36          240.48   
232               9.42                14.13            12          169.56   
413              11.55                17.33             6          103.98   
523               8.87                13.31             2           26.62   

     Volume Sold (Liters)  Volume Sold (Gallons)  Store Category  
19                    9.0                   2.37     Supermarket  
69                   36.0                   9.51     Supermarket  
232                   9.0                   2.37      Food Store  
413                  10.5                   2.77     Supermarket  
523                   0.1                   0.02  Wine & Spirits

In [29]:
hv.loc[hv['Item Description'] == "Athena Ouzo Rum", 'Category'] = 1062500
hv.loc[hv['Item Description'] == "Athena Ouzo Rum", 'Category Name'] = "Spiced Rum"

### White Dog Straight Rye HA and White Dog Wheat HA

In [30]:
hv.loc[hv["Item Description"] == "White Dog Straight Rye HA", "Category Name"] = "American Distilled Spirit Specialty"
hv.loc[hv["Item Description"] == "White Dog Wheat HA", "Category Name"] = "American Distilled Spirit Specialty"

### Dekuyper Mixologist Ginger

In [31]:
hv.loc[(hv["Item Description"] == "Dekuyper Mixologist Ginger")]

Invoice/Item Number        Date  Store Number  \
1436349     INV-26611600001  04/16/2020          2515   

                                Store Name         Address        City  \
1436349  Hy-Vee Food Store #1 / Mason City  2400 4th St SW  Mason City   

         Zip Code                Store Location  County Number      County  \
1436349     50401  POINT (-93.235588 43.148451)             17  CERRO GORD   

         Category Category Name  Vendor Number      Vendor Name  Item Number  \
1436349      <NA>          None             65  Jim Beam Brands       973508   

                   Item Description  Pack  Bottle Volume (ml)  \
1436349  Dekuyper Mixologist Ginger    12               750.0   

         State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
1436349               7.36                11.04            24          264.96   

         Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
1436349                  18.0                   4.75     Food Store

In [32]:
hv.loc[(hv["Item Number"] == 973508)]

Invoice/Item Number        Date  Store Number  \
1436349     INV-26611600001  04/16/2020          2515   

                                Store Name         Address        City  \
1436349  Hy-Vee Food Store #1 / Mason City  2400 4th St SW  Mason City   

         Zip Code                Store Location  County Number      County  \
1436349     50401  POINT (-93.235588 43.148451)             17  CERRO GORD   

         Category Category Name  Vendor Number      Vendor Name  Item Number  \
1436349      <NA>          None             65  Jim Beam Brands       973508   

                   Item Description  Pack  Bottle Volume (ml)  \
1436349  Dekuyper Mixologist Ginger    12               750.0   

         State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
1436349               7.36                11.04            24          264.96   

         Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
1436349                  18.0                   4.75     Food Store

In [33]:
hv[(hv['Item Description'].str.contains('Dekuyper', regex=False)) & (hv["Store Name"] == "Hy-Vee Food Store #1 / Mason City")][["Category", "Category Name"]].mode()

Category      Category Name
0   1081400  American Schnapps

In [34]:
hv.loc[hv["Item Description"] == "Dekuyper Mixologist Ginger", "Category Name"] = "American Schnapps"
hv.loc[hv["Item Description"] == "Dekuyper Mixologist Ginger", "Category"] = 1081400

### Crown Royal w/Glass & Coaster

In [35]:
hv.loc[(hv["Item Description"] == "Crown Royal w/Glass & Coaster")]

Invoice/Item Number        Date  Store Number  \
1483107     INV-01622300001  11/16/2016          2556   

                                    Store Name Address  City  Zip Code  \
1483107  Hy-Vee Wine and Spirits / Estherville    None  None      <NA>   

        Store Location  County Number County  Category Category Name  \
1483107           None           <NA>   None      <NA>          None   

         Vendor Number      Vendor Name  Item Number  \
1483107            260  DIAGEO AMERICAS          115   

                      Item Description  Pack  Bottle Volume (ml)  \
1483107  Crown Royal w/Glass & Coaster     6               750.0   

         State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
1483107              15.07                22.61             6          135.66   

         Volume Sold (Liters)  Volume Sold (Gallons)  Store Category  
1483107                   4.5                   1.19  Wine & Spirits

In [36]:
hv[(hv['Item Description'].str.contains('w/Glass', regex=False))]

Invoice/Item Number        Date  Store Number  \
1297        INV-24073000109  12/20/2019          2624   
25961       INV-24323500011  01/03/2020          2626   
35660       INV-24399000028  01/07/2020          2594   
38450       INV-24459100130  01/09/2020          2515   
39592       INV-24462700009  01/09/2020          2636   
42715       INV-24422600038  01/08/2020          2648   
45706       INV-24494900002  01/10/2020          2591   
49566       INV-24565700131  01/15/2020          2620   
49802       INV-24565700132  01/15/2020          2620   
54043       INV-24567100005  01/15/2020          2517   
65706       INV-24736300006  01/22/2020          2584   
71093       INV-24773000001  01/24/2020          2635   
72529       INV-24773000002  01/24/2020          2635   
85213       INV-40651300004  10/04/2021          2635   
91552       INV-40686800001  10/05/2021          2663   
91731       INV-40687900001  10/05/2021          2663   
94287       INV-40687400004  10/05/2021          2663   
96200       INV-40748300014  10/06/2021          2601   
97294       INV-32377100009  12/02/2020          2601   
97440       INV-32334000030  12/01/2020          2666   
98399       INV-32374700002  12/02/2020          2500   
101027      INV-32370800007  12/02/2020          2565   
101716      INV-32382500076  12/02/2020          2611   
104639      INV-32411400006  12/03/2020          2515   
110907      INV-32513100004  12/08/2020          2604   
111310      INV-32498900004  12/07/2020          2508   
113748      INV-32451400001  12/04/2020          2549   
120434      INV-32635300002  12/11/2020          2550   
120596      INV-32644700001  12/11/2020          2518   
122735      INV-32635900113  12/11/2020          2551   
123454      INV-32605100003  12/10/2020          2515   
124096      INV-32635300001  12/11/2020          2550   
124307      INV-32604700001  12/10/2020          2515   
126434      INV-32710000005  12/15/2020          2666   
127112      INV-32708300083  12/15/2020          2535   
134269      INV-32791400002  12/16/2020          2629   
134376      INV-32791500001  12/16/2020          2629   
135088      INV-32762400007  12/16/2020          2673   
138270      INV-32818200001  12/17/2020          2515   
139615      INV-32851300001  12/18/2020          2602   
142762      INV-32818700002  12/17/2020          2515   
145162      INV-32904800012  12/22/2020          2666   
147327      INV-32904800025  12/22/2020          2666   
155618      INV-32960900009  12/23/2020          2653   
156127      INV-33028300070  12/26/2020          2567   
157060      INV-33029000162  12/26/2020          2539   
158231      INV-33021400093  12/26/2020          2566   
160896      INV-33021400091  12/26/2020          2566   
164750      INV-33061100075  12/28/2020          2669   
165280      INV-33053600004  12/28/2020          2626   
173270      INV-33109000050  12/30/2020          2673   
177416      INV-40753400011  10/06/2021          2558   
178122      INV-40789100049  10/08/2021          2624   
178684      INV-40802800013  10/08/2021          2655   
180380      INV-40762700013  10/07/2021          2572   
181305      INV-40858100022  10/11/2021          2616   
183539      INV-40849900006  10/11/2021          2625   
184132      INV-40884900014  10/12/2021          2544   
184208      INV-40858100016  10/11/2021          2616   
185494      INV-40891900090  10/12/2021          2650   
186680      INV-40900600061  10/12/2021          2604   
188077      INV-40940700012  10/13/2021          2629   
189489      INV-40966100007  10/14/2021          2572   
189806      INV-40962300004  10/14/2021          2594   
192146      INV-41016300004  10/15/2021          2591   
192484      INV-41016300008  10/15/2021          2591   
192699      INV-40994900156  10/15/2021          6209   
194618      INV-41041900011  10/15/2021          2560   
196147      INV-40986600003  10/14/2021          2515   
196516      INV-40

In [37]:
hv.loc[hv["Item Description"] == "Crown Royal w/Glass & Coaster", "Category Name"] = "Temporary & Specialty Packages"
hv.loc[hv["Item Description"] == "Crown Royal w/Glass & Coaster", "Category"] = 1700000

### Cedar Ridge

In [38]:
hv.loc[(hv["Item Description"] == "Cedar Ridge Reserve Bourbon")]

Invoice/Item Number        Date  Store Number  \
1562201     INV-27171500042  05/11/2020          2662   

                                Store Name                Address       City  \
1562201  Hy-Vee Wine & Spirits / Muscatine  522 Mulberry, Suite A  Muscatine   

         Zip Code               Store Location  County Number     County  \
1562201     52761  POINT (-91.04437 41.426615)             70  MUSCATINE   

         Category Category Name  Vendor Number               Vendor Name  \
1562201      <NA>          None            125  CEDAR RIDGE VINEYARDS LL   

         Item Number             Item Description  Pack  Bottle Volume (ml)  \
1562201        27849  Cedar Ridge Reserve Bourbon     6               750.0   

         State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
1562201              25.47                38.21             1           38.21   

         Volume Sold (Liters)  Volume Sold (Gallons)  Store Category  
1562201                  0.75                   0.19  Wine & Spirits

In [39]:
hv[(hv['Item Description'].str.contains('Cedar Ridge', regex=False))].head()

Invoice/Item Number        Date  Store Number  \
102     INV-24115200018  12/23/2019          2630   
169     INV-24073800077  12/20/2019          2524   
441     INV-24073800035  12/20/2019          2524   
558     INV-24081400078  12/20/2019          2551   
566     INV-24115000113  12/23/2019          2619   

                        Store Name          Address             City  \
102      Hy-Vee Drugstore #2 / WDM    1010  60th St  West Des Moines   
169    Hy-Vee Food Store / Dubuque    3500 Dodge St          Dubuque   
441    Hy-Vee Food Store / Dubuque    3500 Dodge St          Dubuque   
558   Hy-Vee Food Store / Chariton  2001 West Court         Chariton   
566  Hy-Vee Wine and Spirits / WDM    1725  74th St  West Des Moines   

     Zip Code                Store Location  County Number   County  Category  \
102     50266  POINT (-93.790534 41.584979)             77     POLK   1011600   
169     52001                          None             31  DUBUQUE   1012300   
441     52001                          None             31  DUBUQUE   1011200   
558     50049                          None             59    LUCAS   1011600   
566     50266  POINT (-93.808855 41.598515)             77     POLK   1011200   

                 Category Name  Vendor Number               Vendor Name  \
102      Straight Rye Whiskies            125  CEDAR RIDGE VINEYARDS LL   
169         Single Malt Scotch            125  CEDAR RIDGE VINEYARDS LL   
441  Straight Bourbon Whiskies            125  CEDAR RIDGE VINEYARDS LL   
558      Straight Rye Whiskies            125  CEDAR RIDGE VINEYARDS LL   
566  Straight Bourbon Whiskies            125  CEDAR RIDGE VINEYARDS LL   

     Item Number                 Item Description  Pack  Bottle Volume (ml)  \
102        27014           Cedar Ridge Malted Rye     6               750.0   
169        27476  Cedar Ridge Single Malt Whiskey     6               750.0   
441        17206              Cedar Ridge Bourbon     6               750.0   
558        27014           Cedar Ridge Malted Rye     6               750.0   
566        17206              Cedar Ridge Bourbon     6               750.0   

     State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
102              19.67                29.51             6          177.06   
169              25.00                37.50             6          225.00   
441              18.10                27.15            12          325.80   
558              19.67                29.51             2           59.02   
566              18.10                27.15             6          162.90   

     Volume Sold (Liters)  Volume Sold (Gallons)  Store Category  
102                   4.5                   1.18       Drugstore  
169                   4.5                   1.18      Food Store  
441                   9.0                   2.37      Food Store  
558                   1.5                   0.39      Food Store  
566                   4.5                   1.18  Wine & Spirits

In [40]:
hv.loc[hv["Item Description"] == "Cedar Ridge Reserve Bourbon", "Category Name"] = "Straight Bourbon Whiskies"
hv.loc[hv["Item Description"] == "Cedar Ridge Reserve Bourbon", "Category"] = 17206

In [41]:
hv.loc[hv["Category"].isna()].head()

Invoice/Item Number        Date  Store Number  \
1580086     INV-27345600003  05/19/2020          2538   

                              Store Name           Address      City  \
1580086  Hy-Vee Food Store #3 / Waterloo  1422 Flammang Dr  Waterloo   

         Zip Code                Store Location  County Number      County  \
1580086     50702  POINT (-92.327917 42.459938)              7  BLACK HAWK   

         Category Category Name  Vendor Number               Vendor Name  \
1580086      <NA>          None            125  CEDAR RIDGE VINEYARDS LL   

         Item Number                                   Item Description  Pack  \
1580086       926786  Cedar Ridge Private Cask Selection Malted Rye ...     6   

         Bottle Volume (ml)  State Bottle Cost  State Bottle Retail  \
1580086               750.0              26.84                40.26   

         Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
1580086           204         8213.04                 153.0   

         Volume Sold (Gallons) Store Category  
1580086                  40.41     Food Store

In [42]:
hv[(hv['Item Description'].str.contains('Cedar Ridge Private Cask', regex=False))].head(10)

Invoice/Item Number        Date  Store Number  \
638964      INV-36984500001  05/26/2021          2647   
1172929     INV-20346900001  07/02/2019          2544   
1220984     INV-28427700001  07/01/2020          2512   
1427527     INV-26574700001  04/15/2020          2648   
1580086     INV-27345600003  05/19/2020          2538   
1584546     INV-27390800001  05/20/2020          2590   
1674650     INV-28239000001  06/24/2020          2670   

                                  Store Name               Address  \
638964              Hy-Vee #7 / Cedar Rapids      5050 Edgewood Rd   
1172929     Hy-Vee Food Store / Marshalltown      802 South Center   
1220984  Hy-Vee Wine and Spirits / Iowa City    1720 Waterfront Dr   
1427527                      Hy-Vee #4 / WDM         555 S 51st St   
1580086      Hy-Vee Food Store #3 / Waterloo      1422 Flammang Dr   
1584546  Hy-Vee Food Store #5 / Cedar Rapids  3235 Oakland Road NE   
1674650       Hy-Vee Food Store / Coralville          2004 8th  St   

                    City  Zip Code                         Store Location  \
638964      Cedar Rapids     52411           POINT (-91.701581 42.030129)   
1172929     Marshalltown     50158           POINT (-92.912817 42.039205)   
1220984        Iowa City     52240   POINT (-91.53046300000001 41.642764)   
1427527  West Des Moines     50265           POINT (-93.773557 41.561197)   
1580086         Waterloo     50702           POINT (-92.327917 42.459938)   
1584546     Cedar Rapids     52402           POINT (-91.658105 42.010971)   
1674650       Coralville     52241  POINT (-91.591868 41.682337000000004)   

         County Number      County  Category     Category Name  Vendor Number  \
638964              57        LINN   1011100  Blended Whiskies            125   
1172929             64    MARSHALL   1011100  Blended Whiskies            125   
1220984             52     JOHNSON   1011100  Blended Whiskies            125   
1427527             77        POLK   1011100  Blended Whiskies            125   
1580086              7  BLACK HAWK      <NA>              None            125   
1584546             57        LINN   1011100  Blended Whiskies            125   
1674650             52     JOHNSON   1011100  Blended Whiskies            125   

                      Vendor Name  Item Number  \
638964   CEDAR RIDGE VINEYARDS LL       928271   
1172929  CEDAR RIDGE VINEYARDS LL       926685   
1220984  CEDAR RIDGE VINEYARDS LL       928270   
1427527  CEDAR RIDGE VINEYARDS LL       928270   
1580086  CEDAR RIDGE VINEYARDS LL       926786   
1584546  CEDAR RIDGE VINEYARDS LL       928271   
1674650  CEDAR RIDGE VINEYARDS LL       928270   

                                          Item Description  Pack  \
638964   Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1172929  Cedar Ridge Private Cask Wheat Whiskey Selecti...     6   
1220984  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1427527  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1580086  Cedar Ridge Private Cask Selection Malted Rye ...     6   
1584546  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1674650  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   

         Bottle Volume (ml)  State Bottle Cost  State Bottle Retail  \
638964                750.0              23.00                34.50   
1172929               750.0              26.84                40.26   
1220984               750.0              21.00                31.50   
1427527               750.0              21.00                31.50   
1580086               750.0              26.84                40.26   
1584546               750.0              23.00                34.50   
1674650               750.0              21.00                31.50   

         Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
638964            216         7452.00                 162.0   
1172929           222         8937.72                 166.5   
1220984           25

In [43]:
hv[(hv['Item Description'].str.contains('Cedar Ridge Private Cask', regex=False))].head(10)

Invoice/Item Number        Date  Store Number  \
638964      INV-36984500001  05/26/2021          2647   
1172929     INV-20346900001  07/02/2019          2544   
1220984     INV-28427700001  07/01/2020          2512   
1427527     INV-26574700001  04/15/2020          2648   
1580086     INV-27345600003  05/19/2020          2538   
1584546     INV-27390800001  05/20/2020          2590   
1674650     INV-28239000001  06/24/2020          2670   

                                  Store Name               Address  \
638964              Hy-Vee #7 / Cedar Rapids      5050 Edgewood Rd   
1172929     Hy-Vee Food Store / Marshalltown      802 South Center   
1220984  Hy-Vee Wine and Spirits / Iowa City    1720 Waterfront Dr   
1427527                      Hy-Vee #4 / WDM         555 S 51st St   
1580086      Hy-Vee Food Store #3 / Waterloo      1422 Flammang Dr   
1584546  Hy-Vee Food Store #5 / Cedar Rapids  3235 Oakland Road NE   
1674650       Hy-Vee Food Store / Coralville          2004 8th  St   

                    City  Zip Code                         Store Location  \
638964      Cedar Rapids     52411           POINT (-91.701581 42.030129)   
1172929     Marshalltown     50158           POINT (-92.912817 42.039205)   
1220984        Iowa City     52240   POINT (-91.53046300000001 41.642764)   
1427527  West Des Moines     50265           POINT (-93.773557 41.561197)   
1580086         Waterloo     50702           POINT (-92.327917 42.459938)   
1584546     Cedar Rapids     52402           POINT (-91.658105 42.010971)   
1674650       Coralville     52241  POINT (-91.591868 41.682337000000004)   

         County Number      County  Category     Category Name  Vendor Number  \
638964              57        LINN   1011100  Blended Whiskies            125   
1172929             64    MARSHALL   1011100  Blended Whiskies            125   
1220984             52     JOHNSON   1011100  Blended Whiskies            125   
1427527             77        POLK   1011100  Blended Whiskies            125   
1580086              7  BLACK HAWK      <NA>              None            125   
1584546             57        LINN   1011100  Blended Whiskies            125   
1674650             52     JOHNSON   1011100  Blended Whiskies            125   

                      Vendor Name  Item Number  \
638964   CEDAR RIDGE VINEYARDS LL       928271   
1172929  CEDAR RIDGE VINEYARDS LL       926685   
1220984  CEDAR RIDGE VINEYARDS LL       928270   
1427527  CEDAR RIDGE VINEYARDS LL       928270   
1580086  CEDAR RIDGE VINEYARDS LL       926786   
1584546  CEDAR RIDGE VINEYARDS LL       928271   
1674650  CEDAR RIDGE VINEYARDS LL       928270   

                                          Item Description  Pack  \
638964   Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1172929  Cedar Ridge Private Cask Wheat Whiskey Selecti...     6   
1220984  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1427527  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1580086  Cedar Ridge Private Cask Selection Malted Rye ...     6   
1584546  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   
1674650  Cedar Ridge Private Cask Iowa Bourbon Selectio...     6   

         Bottle Volume (ml)  State Bottle Cost  State Bottle Retail  \
638964                750.0              23.00                34.50   
1172929               750.0              26.84                40.26   
1220984               750.0              21.00                31.50   
1427527               750.0              21.00                31.50   
1580086               750.0              26.84                40.26   
1584546               750.0              23.00                34.50   
1674650               750.0              21.00                31.50   

         Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
638964            216         7452.00                 162.0   
1172929           222         8937.72                 166.5   
1220984           25

In [44]:
hv[(hv['Item Description'].str.contains('Malted Rye', regex=False))].head()

Invoice/Item Number        Date  Store Number  \
102      INV-24115200018  12/23/2019          2630   
558      INV-24081400078  12/20/2019          2551   
4952     INV-24147600048  12/24/2019          2513   
7049     INV-24145400019  12/24/2019          2575   
8202     INV-24159000089  12/26/2019          2647   

                            Store Name              Address             City  \
102          Hy-Vee Drugstore #2 / WDM        1010  60th St  West Des Moines   
558       Hy-Vee Food Store / Chariton      2001 West Court         Chariton   
4952  Hy-Vee Food Store #2 / Iowa City       812  S 1st Ave        Iowa City   
7049   Hy-Vee Food Store #1 / Waterloo  2834 Ansborough Ave         Waterloo   
8202          Hy-Vee #7 / Cedar Rapids     5050 Edgewood Rd     Cedar Rapids   

      Zip Code                        Store Location  County Number  \
102      50266          POINT (-93.790534 41.584979)             77   
558      50049                                  None             59   
4952     52240                                  None             52   
7049     50702  POINT (-92.37662400000002 42.471767)              7   
8202     52411          POINT (-91.701581 42.030129)             57   

          County  Category          Category Name  Vendor Number  \
102         POLK   1011600  Straight Rye Whiskies            125   
558        LUCAS   1011600  Straight Rye Whiskies            125   
4952     JOHNSON   1011600  Straight Rye Whiskies            125   
7049  BLACK HAWK   1011600  Straight Rye Whiskies            125   
8202        LINN   1011600  Straight Rye Whiskies            125   

                   Vendor Name  Item Number        Item Description  Pack  \
102   CEDAR RIDGE VINEYARDS LL        27014  Cedar Ridge Malted Rye     6   
558   CEDAR RIDGE VINEYARDS LL        27014  Cedar Ridge Malted Rye     6   
4952  CEDAR RIDGE VINEYARDS LL        27014  Cedar Ridge Malted Rye     6   
7049  CEDAR RIDGE VINEYARDS LL        27014  Cedar Ridge Malted Rye     6   
8202  CEDAR RIDGE VINEYARDS LL        26770  Cedar Ridge Malted Rye    12   

      Bottle Volume (ml)  State Bottle Cost  State Bottle Retail  \
102                750.0              19.67                29.51   
558                750.0              19.67                29.51   
4952               750.0              19.67                29.51   
7049               750.0              19.67                29.51   
8202               200.0               6.00                 9.00   

      Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
102              6          177.06                   4.5   
558              2           59.02                   1.5   
4952             6          177.06                   4.5   
7049             6          177.06                   4.5   
8202             3           27.00                   0.6   

      Volume Sold (Gallons) Store Category  
102                    1.18      Drugstore  
558                    0.39     Food Store  
4952                   1.18     Food Store  
7049                   1.18     Food Store  
8202                   0.15    Supermarket

In [45]:
pd.set_option('max_colwidth', 800)

In [46]:
hv[(hv['Category'].isnull())]["Item Description"]

1580086    Cedar Ridge Private Cask Selection Malted Rye 115prf
Name: Item Description, dtype: object

In [47]:
hv.loc[hv["Item Description"] == "Cedar Ridge Private Cask Selection Malted Rye 115prf", "Category Name"] = "Cedar Ridge Malted Rye"
hv.loc[hv["Item Description"] == "Cedar Ridge Private Cask Selection Malted Rye 115prf", "Category"] = 27014

In [48]:
null_counts = hv.isnull().sum()
null_counts

Invoice/Item Number           0
Date                          0
Store Number                  0
Store Name                    0
Address                    1503
City                       1503
Zip Code                   1503
Store Location           266054
County Number              1503
County                     1503
Category                      0
Category Name                 0
Vendor Number                 1
Vendor Name                   1
Item Number                   0
Item Description              0
Pack                          1
Bottle Volume (ml)            1
State Bottle Cost             1
State Bottle Retail           1
Bottles Sold                  1
Sale (Dollars)                1
Volume Sold (Liters)          1
Volume Sold (Gallons)         1
Store Category                0
dtype: int64

## Vendor Number

In [49]:
hv[hv['Vendor Number'].isna()]

Invoice/Item Number        Date  Store Number  \
269636     INV-33728600005  01/25/2021          2633   

                          Store Name          Address        City  Zip Code  \
269636  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines     50320   

                      Store Location  County Number County  Category  \
269636  POINT (-93.596754 41.554101)             77   POLK   1022200   

             Category Name  Vendor Number Vendor Name  Item Number  \
269636  100% Agave Tequila           <NA>        None       988100   

          Item Description  Pack  Bottle Volume (ml)  State Bottle Cost  \
269636  Monte Alban Silver    12               750.0              12.98   

        State Bottle Retail  Bottles Sold  Sale (Dollars)  \
269636                15.56            12          186.72   

        Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
269636                   9.0                   2.37    Supermarket

In [50]:
#checking who sells other products with that item number
hv[hv['Item Number'] == 988100]

Invoice/Item Number        Date  Store Number  \
269636     INV-33728600005  01/25/2021          2633   
343546     INV-41510400006  11/01/2021          2633   
506585     INV-35909800006  04/19/2021          2633   
670913     INV-37056600007  06/01/2021          2633   

                          Store Name          Address        City  Zip Code  \
269636  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines     50320   
343546  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines     50320   
506585  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines     50320   
670913  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines     50320   

                      Store Location  County Number County  Category  \
269636  POINT (-93.596754 41.554101)             77   POLK   1022200   
343546  POINT (-93.596754 41.554101)             77   POLK   1022200   
506585  POINT (-93.596754 41.554101)             77   POLK   1022200   
670913  POINT (-93.596754 41.554101)             77   POLK   1022200   

             Category Name  Vendor Number            Vendor Name  Item Number  \
269636  100% Agave Tequila           <NA>                   None       988100   
343546  100% Agave Tequila            421   SAZERAC COMPANY  INC       988100   
506585  100% Agave Tequila             55  SAZERAC NORTH AMERICA       988100   
670913  100% Agave Tequila             55  SAZERAC NORTH AMERICA       988100   

          Item Description  Pack  Bottle Volume (ml)  State Bottle Cost  \
269636  Monte Alban Silver    12               750.0              12.98   
343546  Monte Alban Silver    12               750.0              10.37   
506585  Monte Alban Silver    12               750.0              10.37   
670913  Monte Alban Silver    12               750.0              10.37   

        State Bottle Retail  Bottles Sold  Sale (Dollars)  \
269636                15.56            12          186.72   
343546                15.56            12          186.72   
506585                15.56            12          186.72   
670913                15.56            12          186.72   

        Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
269636                   9.0                   2.37    Supermarket  
343546                   9.0                   2.37    Supermarket  
506585                   9.0                   2.37    Supermarket  
670913                   9.0                   2.37    Supermarket

In [51]:
hv.loc[(hv["Item Description"] == "Monte Alban Silver") & (hv["Vendor Name"].isna()), "Vendor Name"] = "SAZERAC COMPANY INC"
hv.loc[(hv["Item Description"] == "Monte Alban Silver") & (hv["Vendor Number"].isna()), "Vendor Number"] = 55

In [52]:
hv[hv['Vendor Number'].isna()]

Empty DataFrame
Columns: [Invoice/Item Number, Date, Store Number, Store Name, Address, City, Zip Code, Store Location, County Number, County, Category, Category Name, Vendor Number, Vendor Name, Item Number, Item Description, Pack, Bottle Volume (ml), State Bottle Cost, State Bottle Retail, Bottles Sold, Sale (Dollars), Volume Sold (Liters), Volume Sold (Gallons), Store Category]
Index: []

### INV-30475900020

In [53]:
hv[hv['Pack'].isna()]

Invoice/Item Number        Date  Store Number  \
1745324     INV-30475900020  09/23/2020          2555   

                         Store Name    Address    City  Zip Code  \
1745324  Hy-Vee Food Store / Keokuk  3111 Main  Keokuk     52632   

                      Store Location  County Number County  Category  \
1745324  POINT (-91.403752 40.41538)             56    LEE   1041100   

             Category Name  Vendor Number Vendor Name  Item Number  \
1745324  American Dry Gins            434   LUXCO INC        31658   

        Item Description  Pack  Bottle Volume (ml)  State Bottle Cost  \
1745324         Paramoun  <NA>                 NaN                NaN   

         State Bottle Retail  Bottles Sold  Sale (Dollars)  \
1745324                  NaN          <NA>             NaN   

         Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
1745324                   NaN                    NaN     Food Store

In [54]:
#checking who sells other products with that item number
hv[hv['Item Number'] == 31658].head()

Invoice/Item Number        Date  Store Number  \
32       INV-24112400010  12/23/2019          2568   
135      INV-24073800087  12/20/2019          2524   
645      INV-24070400065  12/20/2019          2569   
1276     INV-24109000012  12/23/2019          2572   
3233     INV-24066600026  12/20/2019          2553   

                               Store Name               Address          City  \
32    Hy-Vee Food Store #4 / Cedar Rapids  1556 First Avenue NE  Cedar Rapids   
135           Hy-Vee Food Store / Dubuque         3500 Dodge St       Dubuque   
645   Hy-Vee Food Store #2 / Cedar Rapids   279 Collins Road NE  Cedar Rapids   
1276      Hy-Vee Food Store / Cedar Falls       6301 University   Cedar Falls   
3233          Hy-Vee Food Store / Creston           600 Sheldon       Creston   

      Zip Code                Store Location  County Number      County  \
32       52402  POINT (-91.650755 41.988967)             57        LINN   
135      52001                          None             31     DUBUQUE   
645      52402  POINT (-91.630677 42.027805)             57        LINN   
1276     50613  POINT (-92.435236 42.512789)              7  BLACK HAWK   
3233     50801  POINT (-94.372552 41.051462)             88       UNION   

      Category      Category Name  Vendor Number Vendor Name  Item Number  \
32     1041100  American Dry Gins            434   LUXCO INC        31658   
135    1041100  American Dry Gins            434   LUXCO INC        31658   
645    1041100  American Dry Gins            434   LUXCO INC        31658   
1276   1041100  American Dry Gins            434   LUXCO INC        31658   
3233   1041100  American Dry Gins            434   LUXCO INC        31658   

     Item Description  Pack  Bottle Volume (ml)  State Bottle Cost  \
32      Paramount Gin     6              1750.0               7.11   
135     Paramount Gin     6              1750.0               7.11   
645     Paramount Gin     6              1750.0               7.11   
1276    Paramount Gin     6              1750.0               7.11   
3233    Paramount Gin     6              1750.0               7.11   

      State Bottle Retail  Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
32                  10.67             6           64.02                  10.5   
135                 10.67             6           64.02                  10.5   
645                 10.67             6           64.02                  10.5   
1276                10.67             6           64.02                  10.5   
3233                10.67             6           64.02                  10.5   

      Volume Sold (Gallons) Store Category  
32                     2.77     Food Store  
135                    2.77     Food Store  
645                    2.77     Food Store  
1276                   2.77     Food Store  
3233                   2.77     Food Store

In [55]:
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["Pack"].isna()), "Pack"] = 6
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["Bottle Volume (ml)"].isna()), "Bottle Volume (ml)"] = 1750.0
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["State Bottle Cost"].isna()), "State Bottle Cost"] = 7.11
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["State Bottle Retail"].isna()), "State Bottle Retail"] = 10.67
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["Bottles Sold"].isna()), "Bottles Sold"] = 6
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["Sale (Dollars)"].isna()), "Sale (Dollars)"] = 64.02
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["Volume Sold (Liters)"].isna()), "Volume Sold (Liters)"] = 10.5
hv.loc[(hv["Item Description"] == "Paramoun") & (hv["Volume Sold (Gallons)"].isna()), "Volume Sold (Gallons)"] = 2.77

### Address

In [56]:
null_counts = hv.isnull().sum()
null_counts

Invoice/Item Number           0
Date                          0
Store Number                  0
Store Name                    0
Address                    1503
City                       1503
Zip Code                   1503
Store Location           266054
County Number              1503
County                     1503
Category                      0
Category Name                 0
Vendor Number                 0
Vendor Name                   0
Item Number                   0
Item Description              0
Pack                          0
Bottle Volume (ml)            0
State Bottle Cost             0
State Bottle Retail           0
Bottles Sold                  0
Sale (Dollars)                0
Volume Sold (Liters)          0
Volume Sold (Gallons)         0
Store Category                0
dtype: int64

#### Algona

In [57]:
hv[hv['Address'].isna()].head()

Invoice/Item Number        Date  Store Number  \
2478     INV-24164900115  12/26/2019          2585   
4294     INV-24164900174  12/26/2019          2585   
4297     INV-24164900172  12/26/2019          2585   
4323     INV-24164900204  12/26/2019          2585   
4330     INV-24164900169  12/26/2019          2585   

                            Store Name Address  City  Zip Code Store Location  \
2478  Hy-Vee Wine and Spirits / Algona    None  None      <NA>           None   
4294  Hy-Vee Wine and Spirits / Algona    None  None      <NA>           None   
4297  Hy-Vee Wine and Spirits / Algona    None  None      <NA>           None   
4323  Hy-Vee Wine and Spirits / Algona    None  None      <NA>           None   
4330  Hy-Vee Wine and Spirits / Algona    None  None      <NA>           None   

      County Number County  Category       Category Name  Vendor Number  \
2478           <NA>   None   1081400   American Schnapps             65   
4294           <NA>   None   1011400  Tennessee Whiskies             85   
4297           <NA>   None   1052100   Imported Brandies            389   
4323           <NA>   None   1051100   American Brandies            434   
4330           <NA>   None   1052100   Imported Brandies            389   

                 Vendor Name  Item Number              Item Description  Pack  \
2478         Jim Beam Brands        82780      Dekuyper Buttershots PET    12   
4294      Brown Forman Corp.        26589                Gentleman Jack     6   
4297  REMY COINTREAU USA INC        49186              Remy Martin VSOP    12   
4323               LUXCO INC        54056          Arrow Apricot Brandy    12   
4330  REMY COINTREAU USA INC        50686  St Remy VSOP Napoleon Brandy    12   

      Bottle Volume (ml)  State Bottle Cost  State Bottle Retail  \
2478               750.0               6.50                 9.75   
4294              1750.0              35.24                52.86   
4297               750.0              23.98                35.97   
4323               750.0               5.38                 8.07   
4330               750.0               7.24                10.86   

      Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
2478            12          117.00                   9.0   
4294             2          105.72                   3.5   
4297             6          215.82                   4.5   
4323             6           48.42                   4.5   
4330             6           65.16                   4.5   

      Volume Sold (Gallons)  Store Category  
2478                   2.37  Wine & Spirits  
4294                   0.92  Wine & Spirits  
4297                   1.18  Wine & Spirits  
4323                   1.18  Wine & Spirits  
4330                   1.18  Wine & Spirits

In [58]:
hv[(hv['Store Name'].str.contains('Algona', regex=False)) & (hv["Address"].notna())].head()

Invoice/Item Number        Date  Store Number  \
25297     INV-24300600053  01/02/2020          2585   
25332     INV-24300600056  01/02/2020          2585   
25336     INV-24300600070  01/02/2020          2585   
25710     INV-24300600014  01/02/2020          2585   
25728     INV-24300600093  01/02/2020          2585   

                             Store Name                     Address    City  \
25297  Hy-Vee Wine and Spirits / Algona  1502 Highway 169 N Suite 1  Algona   
25332  Hy-Vee Wine and Spirits / Algona  1502 Highway 169 N Suite 1  Algona   
25336  Hy-Vee Wine and Spirits / Algona  1502 Highway 169 N Suite 1  Algona   
25710  Hy-Vee Wine and Spirits / Algona  1502 Highway 169 N Suite 1  Algona   
25728  Hy-Vee Wine and Spirits / Algona  1502 Highway 169 N Suite 1  Algona   

       Zip Code               Store Location  County Number   County  \
25297     50511  POINT (-94.235955 43.08091)             55  KOSSUTH   
25332     50511  POINT (-94.235955 43.08091)             55  KOSSUTH   
25336     50511  POINT (-94.235955 43.08091)             55  KOSSUTH   
25710     50511  POINT (-94.235955 43.08091)             55  KOSSUTH   
25728     50511  POINT (-94.235955 43.08091)             55  KOSSUTH   

       Category                Category Name  Vendor Number  \
25297   1031200      American Flavored Vodka            259   
25332   1081300  American Cordials & Liqueur            259   
25336   1012100            Canadian Whiskies            380   
25710   1012100            Canadian Whiskies            421   
25728   1081400            American Schnapps            380   

                Vendor Name  Item Number              Item Description  Pack  \
25297    Heaven Hill Brands        39922               Deep Eddy Peach    12   
25332    Heaven Hill Brands        72722          Christian Bros Honey    12   
25336     Phillips Beverage        14484     Revel Stoke Roasted Apple    12   
25710  SAZERAC COMPANY  INC        11346  Seagrams VO Canadian Whiskey    12   
25728     Phillips Beverage        84617   Phillips Root Beer Schnapps    12   

       Bottle Volume (ml)  State Bottle Cost  State Bottle Retail  \
25297               750.0               8.50                12.75   
25332               750.0               6.00                 9.00   
25336               750.0               8.34                12.51   
25710               750.0               5.23                 7.85   
25728              1000.0               5.50                 8.25   

       Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
25297             6           76.50                   4.5   
25332             6           54.00                   4.5   
25336             6           71.94                   4.5   
25710            12           94.20                   9.0   
25728             6           49.50                   6.0   

       Volume Sold (Gallons)  Store Category  
25297                   1.18  Wine & Spirits  
25332                   1.18  Wine & Spirits  
25336                   1.18  Wine & Spirits  
25710                   2.37  Wine & Spirits  
25728                   1.58  Wine & Spirits

In [ ]:
missing_store_address = hv[hv["Address"].isna()]["Store Name"].dropna().unique()

In [59]:
missing_store_address = hv[hv["Address"].isna()]["Store Name"].dropna().unique()
missing_store_city = hv[hv["City"].isna()]["Store Name"].dropna().unique()
missing_store_zip = hv[hv["Zip Code"].isna()]["Store Name"].dropna().unique()
missing_store_location = hv[hv["Store Location"].isna()]["Store Name"].dropna().unique()
missing_store_county_num = hv[hv["County Number"].isna()]["Store Name"].dropna().unique()
missing_store_county = hv[hv["County"].isna()]["Store Name"].dropna().unique()

In [60]:
missing_store_address

array(['Hy-Vee Wine and Spirits / Algona',
       'Hy-Vee Food Store / Mount Ayr',
       'Hy-Vee Food Store #2 / State Ankeny',
       'Hy-Vee Food Store / Corning', 'Hy-Vee Wine and Spirits / Lemars',
       'Hy-Vee Food Store / Centerville',
       'Hy-Vee Wine & Spirits #1 / MLK',
       'Hy-Vee Food and Drug #6 / Cedar Rapids',
       'Hy-Vee Food Store #1 / Ottumwa',
       'Hy-Vee Wine and Spirits / Estherville',
       'Hy-Vee Wine and Spirits / Atlantic',
       'Hy-Vee Drugstore / Council Bluffs'], dtype=object)

In [61]:
for store_name in missing_store_address:
    # Grab mode for `Address` using `Store Name`
    address_from_mode = hv.loc[(hv["Store Name"] == store_name)]["Address"].mode()

    # For now skipping none mode values
    if len(address_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    address_from_mode = address_from_mode[0]
    
    # Find where `Address` is missing ("N/A" or "None")
    #    Then replace with new value from mode
    hv.loc[hv['Store Name'] == store_name, 'Address'] = hv.loc[hv['Store Name'] == store_name, 'Address'].fillna(address_from_mode)

    
for store_name in missing_store_city:
    # Grab mode for `City` using `Store Name`
    city_from_mode = hv.loc[(hv["Store Name"] == store_name)]["City"].mode()

    # For now skipping none mode values
    if len(city_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    city_from_mode = city_from_mode[0]
    
    # Find where `City` is missing ("N/A" or "None")
    #    Then replace with new `Category` from mode
    hv.loc[hv['Store Name'] == store_name, 'City'] = hv.loc[hv['Store Name'] == store_name, 'City'].fillna(city_from_mode)

    
for store_name in missing_store_zip:
    # Grab mode for `Address` using `Store Name`
    zip_from_mode = hv.loc[(hv["Store Name"] == store_name)]["Zip Code"].mode()

    # For now skipping none mode values
    if len(zip_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    zip_from_mode = zip_from_mode[0]
    
    # Find where `Address` is missing ("N/A" or "None")
    #    Then replace with new `Category` from mode
    hv.loc[hv['Store Name'] == store_name, 'Zip Code'] = hv.loc[hv['Store Name'] == store_name, 'Zip Code'].fillna(zip_from_mode)

    
for store_name in missing_store_location:
    # Grab mode for `Address` using `Store Name`
    location_from_mode = hv.loc[(hv["Store Name"] == store_name)]["Store Location"].mode()

    # For now skipping none mode values
    if len(location_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    location_from_mode = location_from_mode[0]
    
    # Find where `Address` is missing ("N/A" or "None")
    #    Then replace with new `Category` from mode
    hv.loc[hv['Store Name'] == store_name, 'Store Location'] = hv.loc[hv['Store Name'] == store_name, 'Store Location'].fillna(location_from_mode)

    
for store_name in missing_store_county_num:
    # Grab mode for `Address` using `Store Name`
    county_num_from_mode = hv.loc[(hv["Store Name"] == store_name)]["County Number"].mode()

    # For now skipping none mode values
    if len(county_num_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    county_num_from_mode = county_num_from_mode[0]
    
    # Find where `Address` is missing ("N/A" or "None")
    #    Then replace with new `Category` from mode
    hv.loc[hv['Store Name'] == store_name, 'County Number'] = hv.loc[hv['Store Name'] == store_name, 'County Number'].fillna(county_num_from_mode)

    
for store_name in missing_store_county:
    # Grab mode for `Address` using `Store Name`
    county_from_mode = hv.loc[(hv["Store Name"] == store_name)]["County"].mode()

    # For now skipping none mode values
    if len(county_from_mode) == 0:
        continue
        
    # If it's not empty grab first value (mode)
    county_from_mode = county_from_mode[0]
    
    # Find where `Address` is missing ("N/A" or "None")
    #    Then replace with new `Category` from mode
    hv.loc[hv['Store Name'] == store_name, 'County'] = hv.loc[hv['Store Name'] == store_name, 'County'].fillna(county_from_mode)

In [62]:
null_counts = hv.isnull().sum()
null_counts

Invoice/Item Number           0
Date                          0
Store Number                  0
Store Name                    0
Address                       0
City                          0
Zip Code                      0
Store Location           174515
County Number                 0
County                        0
Category                      0
Category Name                 0
Vendor Number                 0
Vendor Name                   0
Item Number                   0
Item Description              0
Pack                          0
Bottle Volume (ml)            0
State Bottle Cost             0
State Bottle Retail           0
Bottles Sold                  0
Sale (Dollars)                0
Volume Sold (Liters)          0
Volume Sold (Gallons)         0
Store Category                0
dtype: int64

In [63]:
hv[hv['Store Location'].isna()]["Store Name"].unique()

array(['Hy-Vee Food Store / Indianola', 'Hy-Vee Food Store / Marion',
       'Hy-Vee Food Store / Chariton', 'Hy-Vee Food Store / Knoxville',
       'Hy-Vee Food Store / Dubuque', 'Hy-Vee Food Store #1 / WDM',
       'Hy-Vee / Windsor Heights', 'Hy-Vee Wine and Spirits / Waterloo',
       'Hy-Vee Wine & Spirits #2 / Davenport',
       'Hy-Vee Food Store #2 / Iowa City',
       'Hy-Vee Wine and Spirits / Spirit Lake',
       'Hy-Vee Fast & Fresh/Altoona', 'Hy-Vee C-Store / Fairfeild',
       'Hy-Vee Dollar Fresh / Maquoketa',
       'Hy-Vee Fast and Fresh / Grimes', 'Hy-Vee Gas / Johnston',
       'Hy-Vee C-Store / Fairfield', 'Hy-Vee Gas - Pleasant Hill',
       'Hy-Vee Wine and Spirits / Esthervill',
       'Hy-Vee Wine and Spirits #2', 'Hy-Vee Food Store / iowa Falls',
       'Hy-Vee Wine and Spirits / Spirit Lak',
       'Hy-Vee / Regal Liquors and Video'], dtype=object)

In [64]:
hv[hv['Store Location'].isna()].head()

Invoice/Item Number        Date  Store Number  \
8      INV-24080300084  12/20/2019          2549   
10     INV-24069700078  12/20/2019          2560   
11     INV-24081400061  12/20/2019          2551   
29     INV-24083700038  12/20/2019          2566   
31     INV-24080300010  12/20/2019          2549   

                       Store Name                     Address       City  \
8   Hy-Vee Food Store / Indianola             910 N Jefferson  Indianola   
10     Hy-Vee Food Store / Marion  3600 Business Hwy 151 East     Marion   
11   Hy-Vee Food Store / Chariton             2001 West Court   Chariton   
29  Hy-Vee Food Store / Knoxville         813 N Lincoln Ste 1  Knoxville   
31  Hy-Vee Food Store / Indianola             910 N Jefferson  Indianola   

    Zip Code Store Location  County Number  County  Category  \
8      50125           None             91  WARREN   1022200   
10     52302           None             57    LINN   1022200   
11     50049           None             59   LUCAS   1012100   
29     50138           None             63  MARION   1081200   
31     50125           None             91  WARREN   1011100   

         Category Name  Vendor Number             Vendor Name  Item Number  \
8   100% Agave Tequila            410  Patron Spirits Company        88291   
10  100% Agave Tequila            619         Campari America        87250   
11   Canadian Whiskies            259      Heaven Hill Brands        11786   
29      Cream Liqueurs            330          Gemini Spirits        80457   
31    Blended Whiskies             65         Jim Beam Brands        27544   

         Item Description  Pack  Bottle Volume (ml)  State Bottle Cost  \
8      Patron Silver Mini    10                50.0              20.30   
10       Cabo Wabo Blanco     6               750.0              19.00   
11       Black Velvet PET    12               750.0               5.23   
29    Ryans Cream Liqueur    12              1000.0               7.23   
31  Red Stag Black Cherry    12               750.0              10.50   

    State Bottle Retail  Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
8                 30.45             1           30.45                  0.05   
10                28.50             6          171.00                  4.50   
11                 7.85            12           94.20                  9.00   
29                10.85            12          130.20                 12.00   
31                15.75            24          378.00                 18.00   

    Volume Sold (Gallons) Store Category  
8                    0.01     Food Store  
10                   1.18     Food Store  
11                   2.37     Food Store  
29                   3.17     Food Store  
31                   4.75     Food Store

In [65]:
hv[(hv['Store Name'].str.contains('Indianola', regex=False))]["Store Location"].mode()

0    POINT (-93.558417 41.36897)
dtype: object

In [66]:
hv[(hv['Store Name'].str.contains('Knoxville', regex=False)) & (hv["Store Location"].notna())].head(3)

Invoice/Item Number        Date  Store Number  \
236501     INV-33328700002  01/11/2021          2690   
237181     INV-33328700010  01/11/2021          2690   
237476     INV-33328700005  01/11/2021          2690   

                             Store Name                   Address       City  \
236501  Hy-Vee Fast & Fresh - Knoxville  809 North Lincoln Street  Knoxville   
237181  Hy-Vee Fast & Fresh - Knoxville  809 North Lincoln Street  Knoxville   
237476  Hy-Vee Fast & Fresh - Knoxville  809 North Lincoln Street  Knoxville   

        Zip Code                Store Location  County Number  County  \
236501     50138  POINT (-93.109494 41.325392)             63  MARION   
237181     50138  POINT (-93.109494 41.325392)             63  MARION   
237476     50138  POINT (-93.109494 41.325392)             63  MARION   

        Category       Category Name  Vendor Number               Vendor Name  \
236501   1031100     American Vodkas            300  McCormick Distilling Co.   
237181   1011400  Tennessee Whiskies             85        Brown Forman Corp.   
237476   1062400          Spiced Rum            260           DIAGEO AMERICAS   

        Item Number                     Item Description  Pack  \
236501        36904            McCormick 80prf Vodka PET    24   
237181        26822      Jack Daniels Old #7 Black Label    48   
237476        43331  Captain Morgan Original Spiced Mini    12   

        Bottle Volume (ml)  State Bottle Cost  State Bottle Retail  \
236501               375.0                1.8                  2.7   
237181               100.0                2.0                  3.0   
237476                50.0                4.8                  7.2   

        Bottles Sold  Sale (Dollars)  Volume Sold (Liters)  \
236501           144           388.8                 54.00   
237181            12            36.0                  1.20   
237476             3            21.6                  0.15   

        Volume Sold (Gallons) Store Category  
236501                  14.26   Fast & Fresh  
237181                   0.31   Fast & Fresh  
237476                   0.03   Fast & Fresh

In [67]:
hv[(hv['Store Name'].str.contains('Indianola', regex=False)) & (hv["Store Location"].notna())].head(3)

Invoice/Item Number        Date  Store Number              Store Name  \
133299     INV-32765100003  12/16/2020          2676  Hy-Vee Gas / Indianola   
133896     INV-32765100002  12/16/2020          2676  Hy-Vee Gas / Indianola   
134952     INV-32765100004  12/16/2020          2676  Hy-Vee Gas / Indianola   

                    Address       City  Zip Code               Store Location  \
133299  912 N Jefferson Way  Indianola     50125  POINT (-93.558417 41.36897)   
133896  912 N Jefferson Way  Indianola     50125  POINT (-93.558417 41.36897)   
134952  912 N Jefferson Way  Indianola     50125  POINT (-93.558417 41.36897)   

        County Number  County  Category    Category Name  Vendor Number  \
133299             91  WARREN   1031100  American Vodkas            301   
133896             91  WARREN   1031100  American Vodkas            301   
134952             91  WARREN   1081600  Whiskey Liqueur            421   

                 Vendor Name  Item Number  \
133299  FIFTH GENERATION INC        38194   
133896  FIFTH GENERATION INC        38174   
134952  SAZERAC COMPANY  INC        64858   

                                Item Description  Pack  Bottle Volume (ml)  \
133299                 Titos Handmade Vodka Mini     5                50.0   
133896                      Titos Handmade Vodka    12               375.0   
134952  Fireball Cinnamon Whiskey Mini Dispenser     1                50.0   

        State Bottle Cost  State Bottle Retail  Bottles Sold  Sale (Dollars)  \
133299              12.80                19.20             5           96.00   
133896               5.42                 8.13            12           97.56   
134952              25.80                38.70             3          135.00   

        Volume Sold (Liters)  Volume Sold (Gallons) Store Category  
133299                  0.25                   0.06            Gas  
133896                  4.50                   1.18            Gas  
134952                  0.15                   0.03            Gas

- unable to fill the remaining 174515 values as previous records do not exist
- Since missing rows only make up 0.7% of the data, will drop

In [68]:
hv = hv.dropna()

In [69]:
null_counts = hv.isnull().sum()
null_counts

Invoice/Item Number      0
Date                     0
Store Number             0
Store Name               0
Address                  0
City                     0
Zip Code                 0
Store Location           0
County Number            0
County                   0
Category                 0
Category Name            0
Vendor Number            0
Vendor Name              0
Item Number              0
Item Description         0
Pack                     0
Bottle Volume (ml)       0
State Bottle Cost        0
State Bottle Retail      0
Bottles Sold             0
Sale (Dollars)           0
Volume Sold (Liters)     0
Volume Sold (Gallons)    0
Store Category           0
dtype: int64

In [70]:
hv.to_csv("hv_cleaned.csv")

In [76]:
len(hv['Zip Code'].unique())

112

In [74]:
len(hv['Category Name'].unique())

75